### PR TITLE
RavenDB-11089 related changes

### PR DESCRIPTION
**Pull request by myarichuk**
* Make sure queries, index definitions, MoreLikeThis, Patches and Facets adhere to custom json serializers.
* add convention for property name changes
* relevant tests
(RavenDB-11089)

### CODE REVIEW

File: src/Raven.Client/Documents/Conventions/DocumentConventions.cs [hunk 1]
```diff
@@ -38,6 +38,8 @@ public class DocumentConventions : Client.Conventions
 
         internal static readonly DocumentConventions Default = new DocumentConventions();
 
+        public Func<MemberInfo, string> PropertyNameConverter = info => info.Name;
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Conventions/DocumentConventions.cs:41]: if public then it needs to be frozen after DocumentConventions.Freeze

File: src/Raven.Client/Documents/Indexes/ExpressionStringBuilder.cs [hunk 2]
```diff
@@ -125,8 +125,9 @@ private void Out(string s)
 
         private void OutMember(Expression instance, MemberInfo member, Type exprType)
         {
-            bool isId = false;
-            var name = GetPropertyName(member.Name, exprType);
+            var isId = false;
+            var name = _conventions.PropertyNameConverter(member) ?? GetPropertyName(member.Name, exprType);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Indexes/ExpressionStringBuilder.cs:129]: The convention is not null (or does not return null by default, which means that it changes the old behavior and introduces a breaking change.

File: src/Raven.Client/Extensions/PropertyNameConventionJavaScriptCompilationExtension.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+﻿using System.Diagnostics;
+using System.Linq.Expressions;
+using System.Reflection;
+using Lambda2Js;
+using Raven.Client.Documents.Conventions;
+
+namespace Raven.Client.Extensions
+{
+    public class PropertyNameConventionJSMetadataProvider : JavascriptMetadataProvider
```
Comment (ppekrol) [inline src/Raven.Client/Extensions/PropertyNameConventionJavaScriptCompilationExtension.cs:9]: please make it internal

File: src/Raven.Client/Documents/Indexes/ExpressionStringBuilder.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Client/Documents/Indexes/ExpressionStringBuilder.cs:129]: Good point, let's make it null / noop by default

**Pull request merged**